File: fuzzy.py
import itertools
import heapq

def get_names():
    with open('files') as f:
        lines = f.readlines()

    return lines

def match_all(items, pattern, limit=10):

    results = []
    for item in items:
        results.append(fuzzy(pattern, item))

    results = heapq.nlargest(limit, results)

    formatted_results = [(r[0], format_match(r[1], r[2])) for r in results]

    return formatted_results

def format_match(indices, string, fmt='shell'):

    if fmt == 'shell':
        s = "\033[31;1m" # red; bold
        e = "\033[0m"

    else:
        s = '<b>'
        e = '</b>'

    formatted_str = []
    last_idx = 0
    for idx in indices:
        formatted_str.append('%s%s%s%s' % (string[last_idx:idx], s, string[idx], e))
        last_idx = idx + 1
    formatted_str.append(string[last_idx:])
    formatted_str = ''.join(formatted_str)

    return formatted_str

def fuzzy(pattern, string):

    LAST_WAS_SEPARATOR_BONUS = 5
    LAST_WAS_PATH_SEPARATOR_BONUS = 8
    MATCHED_CHAR_BONUS = 2
    SEPARATOR_MATCH_BONUS = 1
    GAP_PENELTY = 2

    score = 0
    matched_indices = []
    last_str_was_sep = True # start True to match first char in string
    sep_matches = 0
    last_seps = 0
    path_last_seps = 0

    pattern_idx = 0
    str_idx = 0

    SEPARATORS = ('_', ' ', '/')
    PATH_SEPARATORS = ('/', '\\')

    str_len = len(string)
    pattern_len = len(pattern)

    for str_idx in range(str_len):
        pattern_char = pattern[pattern_idx] if pattern_idx < pattern_len else None
        str_char = string[str_idx]

        # pattern_is_sep = pattern_char in SEPARATORS
        str_is_sep = str_char in SEPARATORS
        # print 'checking %s against %s [%s, %s]' % (pattern_char, str_char, pattern_idx, str_idx)
        if pattern_char and pattern_char.lower() == str_char.lower():
            # print '  matched %s' % str_char
            matched_indices.append(str_idx)
            score += MATCHED_CHAR_BONUS
            pattern_idx += 1
            if last_str_was_sep:
                # print 'Last sep: "%s"' % last_str_char
                if last_str_char in PATH_SEPARATORS:
                    score += LAST_WAS_PATH_SEPARATOR_BONUS
                    path_last_seps += 1
                else:
                    score += LAST_WAS_SEPARATOR_BONUS
                    last_seps += 1

        # elif str_is_sep and pattern_is_sep:
        elif str_is_sep and pattern_char in SEPARATORS:
            # print '  matched SEP'
            score += SEPARATOR_MATCH_BONUS
            sep_matches += 1
            pattern_idx += 1

        last_str_was_sep = str_is_sep
        last_str_char = str_char

    run_length = 0
    gaps = 0
    # runs = []
    for last_idx, idx in pairwise(matched_indices):
        if idx - last_idx == 1:
            run_length += 1
        else:
            # runs.append(run_length+1)
            score += (run_length + 1) ** 2 # + 1 because fence posts
            run_length = 0
            gaps += 1

    if run_length > 1:
        score += (run_length + 1) ** 2
        # runs.append(run_length+1)

    score -= gaps * GAP_PENELTY

    # score -= len(string) - len(matched_indices)

    # fmt = format_match(matched_indices, string)

    SCORE_REPORT = False
    if SCORE_REPORT:
        print "Matched Chars:\t%02d" % ( len(matched_indices) * MATCHED_CHAR_BONUS
                )
        print "Sep Matches:\t%02d" % (sep_matches * SEPARATOR_MATCH_BONUS)
        print "Last Seps:\t%02d" % (last_seps * LAST_WAS_SEPARATOR_BONUS)
        print "Path Last Seps:\t%02d" % (path_last_seps * LAST_WAS_PATH_SEPARATOR_BONUS)
        print "Runs:\t\t%02d" % sum(r ** 2 for r in runs)
        print "- Gaps:\t\t%02d" % (gaps * 2)
        # print "- Len:\t\t%02d" % (len(string) - len(matched_indices))
        print score, fmt, '\n'

    # return score, fmt
    return score, matched_indices, string

def pairwise(iterable):
    """ Creates an iterator that iterates the original in (n, n+1) pairs
        i -> (i0, i1), (i1, i2), (i2, i3), ... (iN-1, iN) """
    return itertools.izip(iterable, itertools.islice(iterable, 1, None))

if __name__ == '__main__':
    import sys

    names = [
        "erp/automated_tests/tests/controllers/test_accounts.py",
        "erp/controllers/testing/accounttests.py",
        "erp/controllers/testing/accounts.py",
        "erp/automated_tests/tests/controllers/test_app.py",
        "erp/controllers/testing/tracktests.py",
        "erp/controllers/testing/tracks.py",
        "erp/controllers/testing/testpackages.py",
        "erp/controllers/testing/paperinvoicerecipients.py",
        "erp/controllers/testing/emailinvoicerecipients.py",
        "erp/automated_tests/tests/controllers/test_transactions.py",
        ]

    # names = get_names()

    if len(sys.argv) == 2:
        # names = ["erp/controllers/testing/emailinvoicerecipients.py",
        #          "erp/controllers/testing/accounts.py"]
        pattern = sys.argv[1]
        matches = match_all(names, pattern)
        for match in matches:
            print match[0], match[1]

    elif len(sys.argv) == 3:
        score, string = fuzzy(sys.argv[1], sys.argv[2])
        print score, string
    else:
        print fuzzy('cont test acc', '/erp/controllers/testing/accounts.py')

File: fuzzy.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"log"
	"os"
	"runtime"
	"strings"
	"sync"
)

func fuzzy(pattern string, str string) (score int, matched_indices []int) {
	const LastWasSeparatorBonus = 5
	const LastWasPathSeparatorBonus = 8
	const MatchedCharBonus = 2
	const SeparatorMatchBonus = 1
	const GapPenelty = 2

	score = 0
	matched_indices = make([]int, 0, 100)
	last_str_was_sep := true
	// sep_matches := 0
	// last_seps := 0
	// path_last_seps := 0

	pattern_idx := 0

	// SEPARATORS = '_', ' ', '/'
	// PATH SEPARATORS = '/', '\'

	str_len := len([]rune(str))
	pattern_len := len([]rune(pattern))

	var last_str_char string = ""

	for str_idx := 0; str_idx < str_len; str_idx++ {
		var pattern_char string = ""
		if pattern_idx < pattern_len {
			pattern_char = string(pattern[pattern_idx])
		}
		str_char := string(str[str_idx])

		str_is_sep := false
		if isSep(str_char) {
			str_is_sep = true
		}

		if pattern_char != "" && strings.ToLower(pattern_char) == strings.ToLower(str_char) {
			matched_indices = append(matched_indices, str_idx)
			score += MatchedCharBonus
			pattern_idx += 1

			if last_str_was_sep {
				if isPathSep(last_str_char) {
					score += LastWasPathSeparatorBonus
					// path_last_seps += 1
				} else {
					score += LastWasSeparatorBonus
					// last_seps += 1
				}
			}
		} else if str_is_sep && isSep(pattern_char) {
			score += SeparatorMatchBonus
			// sep_matches += 1
			pattern_idx += 1
		}

		last_str_was_sep = str_is_sep
		last_str_char = str_char
	}

	run_length := 0
	gaps := 0
	last_idx := -1

	for _, idx := range matched_indices {
		if last_idx < 0 {
			last_idx = idx
			continue
		}

		if idx-last_idx == 1 {
			run_length += 1
		} else {
			score += (run_length + 1) * (run_length + 1)
			run_length = 0
			gaps += 1
		}
	}

	if run_length > 1 {
		score += (run_length + 1) * (run_length + 1)
	}

	score -= gaps * GapPenelty

	return score, matched_indices
}

func isSep(char string) bool {
	if char == "/" || char == " " || char == "_" {
		return true
	} else {
		return false
	}
}

func isPathSep(char string) bool {
	if char == "/" || char == "\\" {
		return true
	} else {
		return false
	}
}

func formatMatch(indices []int, str string) string {

	s := "\033[31;1m"
	e := "\033[0m"

	hunks := make([]string, len(indices)+1)

	last_idx := 0
	i := 0

	for _, str_idx := range indices {
		hunks[i] = str[last_idx:str_idx] + s + string(str[str_idx]) + e
		last_idx = str_idx + 1
		i++
	}
	hunks[i] = str[last_idx:len(str)]

	formatted_str := strings.Join(hunks, "")

	return formatted_str
}

type matchResult struct {
	score   int
	indices []int
	str     string
}

// An matchResultHeap is a min-heap of matchResults
type matchResultHeap []matchResult

func (h matchResultHeap) Len() int           { return len(h) }
func (h matchResultHeap) Less(i, j int) bool { return h[i].score < h[j].score }
func (h matchResultHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *matchResultHeap) Push(x interface{}) {
	// Push and Pop use pointer receivers because they modify the slice's length,
	// not just its contents.
	*h = append(*h, x.(matchResult))
}

func (h *matchResultHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func matchAll(items []string, pattern string) []string {
	h := &matchResultHeap{}
	heap.Init(h)
	maxScore := -99999999999
	for _, item := range items {
		score, indices := fuzzy(pattern, item)
		r := matchResult{score: score * -1, indices: indices, str: item}
		heap.Push(h, r)

		if r.score > maxScore {
			maxScore = r.score
		}
	}
	// fmt.Println("max score %d", maxScore)

	results := make([]string, 10)
	for i := 0; i < 10; i++ {
		r := heap.Pop(h).(matchResult)
		// fmt.Println("%+v", r)
		results[i] = formatMatch(r.indices, r.str)
	}

	return results
}

func matchAllN(items []string, pattern string) []string {

	ch := make(chan matchResult, 100)
	done := make(chan bool, 1)

	h := &matchResultHeap{}
	heap.Init(h)

	workers := runtime.NumCPU()

	var wg sync.WaitGroup
	wg.Add(workers)

	var final sync.WaitGroup
	final.Add(1)

	producer := func(items []string, c chan<- matchResult) {
		defer wg.Done()
		for _, item := range items {
			score, indices := fuzzy(pattern, item)
			r := matchResult{score: score * -1, indices: indices, str: item}
			c <- r
		}
	}

	chunk_size := len(items) / workers
	for i := 0; i < workers; i++ {
		start := i * chunk_size
		end := (i + 1) * chunk_size
		if end > len(items) {
			end = len(items)
		}
		go producer(items[start:end], ch)
	}

	results := make([]string, 10)

	drain := func() {
		for {
			select {
			case r := <-ch:
				heap.Push(h, r)
			default:
				return
			}
		}
	}

	go func() {
		for {
			select {
			case r := <-ch:
				heap.Push(h, r)
			case <-done:

				// drain the rest of the content of the channel
				drain()

				for i := 0; i < 10; i++ {
					r := heap.Pop(h).(matchResult)
					results[i] = formatMatch(r.indices, r.str)
				}

				final.Done()
				return
			}
		}
	}()

	go func() {
		wg.Wait()
		done <- true
		close(done)
	}()

	final.Wait()

	return results
}

func loadFile(filename string) []string {

	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	lines := make([]string, 0, 4000)

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	return lines
}

func main() {
	pattern := "cont test acc"
	// input := "/erp/controllers/testing/accounts.py"
	// score, indices := fuzzy("cont test acc", input)
	//
	// formatted := formatMatch(indices, input)
	// fmt.Println(score, formatted)

	corpus := loadFile("files")
	// corpus := []string{
	// 	"erp/automated_tests/tests/controllers/test_accounts.py",
	// 	"erp/controllers/testing/accounttests.py",
	// 	"erp/controllers/testing/accounts.py",
	// 	"erp/automated_tests/tests/controllers/test_app.py",
	// 	"erp/controllers/testing/tracktests.py",
	// 	"erp/controllers/testing/tracks.py",
	// 	"erp/controllers/testing/testpackages.py",
	// 	"erp/controllers/testing/paperinvoicerecipients.py",
	// 	"erp/controllers/testing/emailinvoicerecipients.py",
	// 	"erp/automated_tests/tests/controllers/test_transactions.py",
	// }
	results := matchAllN(corpus, pattern)
	// results := matchAll(corpus, pattern)
	for _, result := range results {
		fmt.Println(result)
	}

}
